google_compute_engine/distro/sles_12/utils.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/python
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities that are distro specific for use on SUSE 12."""
+
+import os
+import subprocess
+
+from google_compute_engine import constants
+from google_compute_engine.distro import utils
+
+
+class Utils(utils.Utils):
+  """Utilities used by Linux guest services on SUSE 12."""
+
+  network_path = constants.LOCALBASE + '/etc/sysconfig/network'
+
+  def EnableNetworkInterfaces(
+      self, interfaces, logger, dhclient_script=None):
+    """Enable the list of network interfaces.
+
+    Args:
+      interfaces: list of string, the output device names to enable.
+      logger: logger object, used to write to SysLog and serial port.
+      dhclient_script: string, the path to a dhclient script used by dhclient.
+    """
+    interfaces_to_up = [i for i in interfaces if i != 'eth0']
+    if interfaces_to_up:
+      logger.info('Enabling the Ethernet interfaces %s.', interfaces_to_up)
+      self._WriteIfcfg(interfaces_to_up, logger)
+      self._Ifup(interfaces_to_up, logger)
+
+  def _WriteIfcfg(self, interfaces, logger):
+    """Write ifcfg files for multi-NIC support.
+
+    Overwrites the files. This allows us to update ifcfg-* in the future.
+    Disable the network setup to override this behavior and customize the
+    configurations.
+
+    Args:
+      interfaces: list of string, the output device names to enable.
+      logger: logger object, used to write to SysLog and serial port.
+    """
+    for interface in interfaces:
+      interface_config = os.path.join(
+          self.network_path, 'ifcfg-%s' % interface)
+      interface_content = [
+          '# Added by Google.',
+          'STARTMODE=hotplug',
+          'BOOTPROTO=dhcp',
+          'DHCLIENT_SET_DEFAULT_ROUTE=yes',
+          'DHCLIENT_ROUTE_PRIORITY=10%s00' % interface,
+          'CLOUD_NETCONFIG_MANAGE=yes',
+          'POST_DOWN_SCRIPT=compat:suse:cloud-netconfig-cleanup',
```

Comment:
@jgleissner given the interfaces are setup in the initialization code would we still run the cleanup code from cloud-net-config? Also we'd need to build a generic cloud-netconfig package that provides this script.

Reply:
@rjschwei cloud-netconfig-cleanup removes routing policies associated with the interface that was brought down. cloud-netconfig creates routing policies to make sure packets are sent out through the interface associated with the source IP address. AFACIS the initialization code in this project doesn't do that, so this script would not be needed, unless cloud-netconfig is intended to be used. Which would make sense actually, as it creates the aforementioned routing policies and also can configure secondary (alias) IP addresses, unless that functionality is intended to be replicated in this initialization code.

Reply:
@jgleissner yes, I guess the question if for @ryanwe and @illfelder to answer, what is the expectation how the routing policies for multiple network interfaces are handeled? Does the GCE DHCP server provide the routing information for the new interface, i.e. the policy is expected to be added by the distribuiton dhcp client? I the answer is yes, then

```
'CLOUD_NETCONFIG_MANAGE=yes',
'POST_DOWN_SCRIPT=compat:suse:cloud-netconfig-cleanup',
```
should be removed. The current implementation of cloud-netconfig and the packages we build extract routing information from the metadata in EC2 and Azure and set up routing, as @jgleissner pointed out.

The code would indicate that the answer to my question is "yes, dhcp client gets the route information from the GCE DHCP server and is expected to set he routes", but I want to double check tha this is correct. Thanks

Reply:
@rjschwei, yes that is correct. I'll remove those lines. I also updated our distribution checks to use `distro`. Thanks!

Reply:
The routing policies set up by cloud-netconfig are actually not based on metadata information, at least not directly. cloud-netconfig creates a routing policy for any IP address (i.e. also the one set by DHCP) it finds on any secondary interfaces. This is useful, as the kernel will pick the default route via eth0 when sending packets not directly destined for the network associated with i.e. eth1, even when their source IP address is on that network. So for instance, if you start an instance with two network interfaces, with both having an external IP associated, you can see that you can ssh into the machine on the one associated with eth0 but not on the eth1 one. Using policy routing amends this, which is why we created cloud-netconfig to automatically do that. So the question is whether the GCE initialization code is going to do this.

The metadata information are used by cloud-netconfig, as mentioned, to configured secondary IP addresses in EC2 and Azure. I thought "alias IP" would be GCE's equivalent, but I misunderstood that. It's actually an additional IP range the GCE will route to the VM, probably intended for use with container hosts. So for cloud-netconfig, should we want to use it on GCE, there's nothing really to do at this point. It could set up routing policies for those ranges, but since alias IP ranges are only supported with single-nic instances, it's not necessary.